### PR TITLE
Provide How-to guide for the Kanto Local Digital Twins

### PR DESCRIPTION
**Pull request by antonia-avramova**
[#51] Provide How-to guide for the Kanto Local Digital Twins

Signed-off-by: Antonia Avramova <antonia.avramova@bosch.io>

### CODE REVIEW

File: web/site/content/docs/how-to-guides/local-digital-twins.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+Local digital twins uses the `/etc/local-digital-twins/config.json` to acquire all the remote communication, identification and
+authentication data to establish the remote connection. Because there is no such remote connection the Local digital twins will work offline. Update it with the following:
+
+
+```json
+{
+  "caCert": "/etc/local-digital-twins/iothub.crt",
+  "logFile": "/var/log/local-digital-twins/local-digital-twins.log",
+  "thingsDb": "/var/lib/local-digital-twins/thing.db",
+  "tenantId": "demo",
```
Comment (e-grigorov) [inline web/site/content/docs/how-to-guides/local-digital-twins.md:54]: Can we align the order with the example from the reference guide?

File: web/site/content/docs/how-to-guides/local-digital-twins.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  "deviceId": "demo:device",
+  "authId": "demo_device",
+  "password": "secret"
+}
+```
+
+Restart the Local Digital Twins service for the changes to take effect:
```
Comment (e-grigorov) [inline web/site/content/docs/how-to-guides/local-digital-twins.md:61]: We have a second restart of the service. I guess that only one can be used.

File: web/site/content/docs/how-to-guides/local-digital-twins.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+* The {{% refn "https://github.com/eclipse-kanto/kanto/blob/main/quickstart/kanto_commands_ldt.py" %}} eclipse kanto client {{% /refn %}}
+
+  Navigate to the `quickstart` folder where the resources from the {{% relrefn "hono" %}} Explore via Eclipse Hono   {{% /relrefn %}} guide are located and execute the following script:
+
+  ```shell
+  wget https://github.com/eclipse-kanto/kanto/raw/main/quickstart/kanto_commands_ldt.py
+  ```
+
+### Configure Local digital twins
```
Comment (e-grigorov) [inline web/site/content/docs/how-to-guides/local-digital-twins.md:43]: This step sounds like a prerequisite. Just a question, what about `Before you begin` section?

File: web/site/content/docs/how-to-guides/local-digital-twins.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+}
+```
+
+Restart the Local Digital Twins service for the changes to take effect:
+
+```shell
+sudo systemctl restart local-digital-twins.service
+```
+
+### Receive the structure of the edge device
```
Comment (e-grigorov) [inline web/site/content/docs/how-to-guides/local-digital-twins.md:67]: The steps  should be as short as possible.

File: web/site/content/docs/how-to-guides/local-digital-twins.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+Now we are ready to request the structure of the application via executing the Eclipse kanto client that requires the Local digital twins tenant (-t) and the device identifier (-d):
```
Comment (e-grigorov) [inline web/site/content/docs/how-to-guides/local-digital-twins.md:69]: `structure of the application` - the document should explore the edge digital twins, not the application structure.

File: web/site/content/docs/how-to-guides/local-digital-twins.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+```shell
+python3 kanto_commands_ldt.py -t demo -d demo:device
+```
+
+### Verify
+
+A command with a similar value will be received in the terminal and status 200.
```
Comment (e-grigorov) [inline web/site/content/docs/how-to-guides/local-digital-twins.md:77]: `A command with...` - it's not a command.

File: quickstart/kanto_commands_ldt.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+# terms of the Eclipse Public License 2.0 which is available at
+# https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+# which is available at https://www.apache.org/licenses/LICENSE-2.0.
+#
+# SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+
+import getopt
+import json
+import os
+# import signal
```
Comment (e-grigorov) [inline quickstart/kanto_commands_ldt.py:16]: Clean up such unused imports.

File: quickstart/kanto_commands_ldt.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+import sys
+# import threading
+import time
+import uuid
+from string import Template
+import paho.mqtt.client as mqtt
+
+ditto_receive_thing_template = Template("""
```
Comment (e-grigorov) [inline quickstart/kanto_commands_ldt.py:24]: `receive_thing` - I guess that it should be: `retrieve_things` or just a more generic naming as it's used by other how-to guides.

File: quickstart/kanto_commands_ldt.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    "value": {
+			"thingIds": [
+				"$namespace:$name",
+				"$namespace:$name:edge:containers"
+			]
+		}
+}
+""")
+
+class MQTTClientClass(mqtt.Client):
```
Comment (e-grigorov) [inline quickstart/kanto_commands_ldt.py:42]: Unusual class name.

File: quickstart/kanto_commands_ldt.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+    def on_connect(self, mqttc, obj, flags, rc):
+        print('[client connected]')
+        self.subscribe("command/#")
```
Comment (e-grigorov) [inline quickstart/kanto_commands_ldt.py:46]: I would suggest to stick to [the Hono defined topics](https://www.eclipse.org/hono/docs/user-guide/mqtt-adapter/).

File: quickstart/kanto_commands_ldt.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        self.connect("localhost")
+
+        rc = 0
+        while rc == 0:
+            rc = self.loop_start()
+            namespaced_id = device_id.split(':', 1)
+            payload = ditto_receive_thing_template.substitute(
+                namespace=namespaced_id[0], name=namespaced_id[1],
+                tenant_id=tenant_id, correlation_id=str(uuid.uuid4()))
+            pub_topic = 'e/{}/{}'.format(tenant_id, device_id)
```
Comment (e-grigorov) [inline quickstart/kanto_commands_ldt.py:63]: Use short or long topics, don't mix them.

File: quickstart/kanto_commands_ldt.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+            self.publish(pub_topic, payload)
+            time.sleep(1)
+            rc = self.loop_stop()
+            return rc
```
Comment (e-grigorov) [inline quickstart/kanto_commands_ldt.py:67]: This  return will terminate the loop, how is rc used in the while loop?

File: quickstart/kanto_commands_ldt.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+device_id = os.environ.get("DEVICE_ID") or opts_dict['-d']
+
+uri = 'localhost'
+print('[starting] local-digital-twins commands app for tenant [{}], device [{}] at [{}]'.format(
+    tenant_id, device_id, uri))
+
+
+mqttc = MQTTClientClass()
+mqttc.run()
+# time.sleep(1)
```
Comment (e-grigorov) [inline quickstart/kanto_commands_ldt.py:82]: Clean up

File: quickstart/kanto_commands_ldt.py [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+            return rc
+            
+# Parse command line args
+options, reminder = getopt.getopt(sys.argv[1:], 't:d:')
+opts_dict = dict(options)
+tenant_id = os.environ.get("TENANT") or opts_dict['-t']
+device_id = os.environ.get("DEVICE_ID") or opts_dict['-d']
+
+uri = 'localhost'
+print('[starting] local-digital-twins commands app for tenant [{}], device [{}] at [{}]'.format(
```
Comment (e-grigorov) [inline quickstart/kanto_commands_ldt.py:76]: Align the message with other how-to guides.

File: quickstart/kanto_commands_ldt.py [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+    tenant_id, device_id, uri))
+
+
+mqttc = MQTTClientClass()
+mqttc.run()
+# time.sleep(1)
+print('[stopped] local-digital-twins commands app for tenant [{}], device [{}] at [{}]'.format(
+        tenant_id, device_id, uri))
+exit(0)
```
Comment (e-grigorov) [inline quickstart/kanto_commands_ldt.py:85]: Signal handling is missing.

**Review by e-grigorov — CHANGES_REQUESTED** [review]

**Review by k-gostev — CHANGES_REQUESTED** [review]

File: quickstart/hono_commands_ldt.py [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+            rc = self.loop_start()
+            namespaced_id = device_id.split(':', 1)
+            payload = retrieve_things_template.substitute(
+                namespace=namespaced_id[0], name=namespaced_id[1],
+                tenant_id=tenant_id, correlation_id=str(uuid.uuid4()))
+            pub_topic = 'event/{}/{}'.format(tenant_id, device_id)
+            print(payload)
+            self.publish(pub_topic, payload)
+            time.sleep(1)
+            rc = self.loop_stop()
```
Comment (dimitar-dimitrow) [inline quickstart/hono_commands_ldt.py:67]: I have run a dozen of times the app and even with those timeouts there is a sync issue - no retrieve-response is received as the subscription is not done on time.

**Review by dimitar-dimitrow — CHANGES_REQUESTED** [review]

**Review by dimitar-dimitrow — APPROVED** [review]

File: web/site/content/docs/how-to-guides/offline-edge-device.md [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+
+### Verify
+
+On the shell there will be output of the structure of the edge digital twins with all it's features and properties. Things with following identifiers will be presented:
+* demo:device
+* demo:device:edge:containers
+
+### Clean up
+
+Stop local digital twins service and start suite connector service by executing:
```
Comment (k-gostev) [inline web/site/content/docs/how-to-guides/offline-edge-device.md:87]:
```suggestion
Stop the local digital twins service and start suite connector service by executing:
```

**Review by k-gostev — CHANGES_REQUESTED** [review]

**Review by hristobojilov — APPROVED** [review]

**Review by hristobojilov — APPROVED** [review]

**Review by k-gostev — APPROVED** [review]

**Pull request merged**